ouroboros-network/src/Ouroboros/Network/Connections/Concurrent.hs

```diff
@@ -6,138 +6,215 @@
 {-# LANGUAGE GADTs #-}
 
 module Ouroboros.Network.Connections.Concurrent
-  ( concurrent
+  ( Handler (..)
+  , concurrent
   ) where
 
 import Control.Concurrent.Async
 import Control.Concurrent.MVar
-import Control.Exception (assert, bracket, finally, mask_)
+import Control.Exception (finally, mask, onException)
 import Control.Monad (forM_)
 import Data.Map.Strict (Map)
 import qualified Data.Map.Strict as Map
-import Network.Socket (Socket, SockAddr)
+import Data.Word (Word32)
 
 import Ouroboros.Network.Connections.Types
 
-data Reject (p :: Provenance) = Duplicate
+data Reject reason (p :: Provenance) where
+  -- | A remote connection may be rejected if there is already a connection
+  -- established at its identifier.
+  -- Locally-initiated connections cannot fail in this way, because the
+  -- connection manager will use the existing remotely-initiated one.
+  --
+  -- For TCP sockets, and connection identified by local and remote
+  -- socket addresses, this should never happen (TCP spec sorts out crossed
+  -- connections).
+  --
+  -- Indeed, the choice of connection identifier and resource type should be
+  -- done in such a way that duplicates do not arise, i.e. the identifier
+  -- actually does uniquely identify the resource.
+  Duplicate      :: Reject reason Incoming
+  -- | Rejected according to the domain-specific handler routine.
+  DomainSpecific :: reason -> Reject reason any
 
-data Accept (p :: Provenance) = Accept
+data Accept handle (p :: Provenance) = Accept handle
 
-data Connection
-  = Remote (Async ())
-  | Local (Async ())
-  -- | First is remote, second is local
-  | Both (Async ()) (Async ())
+-- | An action to run for each connection, and a handle giving an interface
+-- to that action, probably using STM for synchronization.
+-- The connection for which this handler is spawned will be closed whenever
+-- the action returns, exceptionally or not.
+data Handler handle = Handler
+  { handle :: !handle
+  , action :: !(IO ())
```

Comment:
Does this have to be in `IO`? Mux can run in `io-sim` and with Snockets on the way I think we should be able to test this in sim.

Reply:
No, I will port it all over to io-sim stuff. I'm just a bit annoyed that there is no `MVar` equivalent in there, so I have to simulate it using `TMVar`s.